File: Programmers/[17679] 프렌즈4블록/Johoseong/17679.py
def solution(m, n, board):
    answer = 0
    field = [list(i) for i in board]

    while True:
        flag = 0
        remove = []

        for i in range(m - 1): # 2*2 같은 블록 찾음 (나중에 지워야해서 위치 기억해둠)
            for j in range(n - 1):
                if field[i][j] == 0:
                    continue

                if field[i][j] == field[i][j + 1] == field[i + 1][j] == field[i + 1][j + 1]:
                    flag = 1
                    remove.append([i, j])
                    remove.append([i, j + 1])
                    remove.append([i + 1, j])
                    remove.append([i + 1, j + 1]) 

        if flag == 0: # 지워지는 거 없음 -> 게임 종료
            break

        while remove: # 지움
            r, c = remove.pop()
            field[r][c] = 0

        for i in range(m - 1): # 위에 있는거 내림
            for j in range(n):
                if field[i + 1][j] == 0 and field[i][j] != 0:
                    field[i + 1][j] = field[i][j]
                    field[i][j] = 0

                    for k in range(i - 1, 0, -1):
                        field[k + 1][j] = field[k][j]
                        field[k][j] = 0
        # print(field)
                
    for i in range(m): # 터진거 개수 셈
        for j in range(n):
            if field[i][j] == 0:
                answer += 1     

    return answer

File: Programmers/[17679] 프렌즈4블록/ONE/17679.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int m, int n, String[] board) {
        int answer = 0;
        char[][] b = new char[m][n];

        for(int i = 0; i < m; i++)
            for(int j = 0; j < n; j++)
                b[i][j] = board[i].charAt(j);

        while (true) {
            boolean check = false;
            boolean[][] remove = new boolean[m][n];

            for (int i = 0; i < m - 1; i++)
                for (int j = 0; j < n - 1; j++)
                    if(b[i][j] != '-' && isSquare(b, i, j)) {
                        remove[i][j] = true;
                        remove[i][j + 1] = true;
                        remove[i + 1][j] = true;
                        remove[i + 1][j + 1] = true;
                        check = true;
                    }

            if(!check)
                break;

            for (int i = 0; i < m; i++)
                for (int j = 0; j < n; j++)
                    if(remove[i][j]) {
                        b[i][j] = '-';
                        answer++;
                    }

            for (int j = 0; j < n; j++) {
                Queue<Character> queue = new LinkedList<>();

                for (int i = m - 1; i >= 0; i--)
                    if(b[i][j] != '-')
                        queue.add(b[i][j]);

                int index = m - 1;
                while (!queue.isEmpty())
                    b[index--][j] = queue.poll();

                for (int i = index; i >= 0; i--)
                    b[i][j] = '-';
            }
        }
        return answer;
    }

    private boolean isSquare(char[][] b, int x, int y) {
        return b[x][y + 1] == b[x][y] && b[x + 1][y] == b[x][y] && b[x + 1][y + 1] == b[x][y];
    }
}